File: Computer Programming/C++/Hashing/Prefix Sum/Subarray Sums Divisible by K.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) 
    {
        int n = nums.size();

        vector<int>prefix(n);
        prefix[0] = nums[0];
        unordered_map<int,int>mp;
        mp[0] = 1;
        for(int i = 1;i<n;i++)
        {
            prefix[i] = prefix[i-1] + nums[i];
        }
        int count = 0;
        for(int i = 0;i<n;i++)
        {
            int mod = ((prefix[i] % k) + k) % k;
            if(mp.count(mod))
            {
                count += mp[mod];
            }
            mp[mod] ++;
        }

        return count;
        



    }
};

File: Computer Programming/Python/Hashing/Subarray Sums Divisible by K.py
class Solution(object):
    def subarraysDivByK(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        
        mp = {}
        prefix = {}
        prefix[0] = nums[0]
        n = len(nums)
        
        for i in range(1,n):
            prefix[i] = prefix[i-1] + nums[i]
        
        mp[0] = 1
        count = 0
        for i in range(n):
            mod = ((prefix[i] % k) + k) % k

            if mod in mp:
                count += mp[mod]
            
            if mod in mp:
                mp[mod] += 1
            else:
                mp[mod] = 1
        
        return count


